### PR TITLE
Create FRSky frontend/backend split

### PR DESCRIPTION
**Pull request by peterbarker**
- will allow for easy addition of e.g. `HAL_FRSKY_D_ENABLE`, for trimming build size (720 bytes)
 - shows clearly data being labelled as for use in sport but also being used for D
 - hides lots of state from different backends
 - seems to show that we only support SPort-passthrough when using external data (rather than also allowing SPort)
 - clears out AP_Frsky_Telem; bidirectional mixes in poorly...

- [X] SITL tests pass
- [X] SPortPassthrough
- [X] SPort
- [ ] D
- [X] SPortPassthrough via FPort

### CODE REVIEW

File: libraries/AP_Frsky_Telem/AP_Frsky_SPort.cpp [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    }
+#endif
+    _port->write(buf2, len);
+}
+
+/*
+ * prepare value for transmission through FrSky link
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
+ */
+uint16_t AP_Frsky_SPort::prep_number(int32_t number, uint8_t digits, uint8_t power)
```
Comment (yaapu) [inline libraries/AP_Frsky_Telem/AP_Frsky_SPort.cpp:193]: this is only used by passthrough so should probably go in the relevant backend.
As a side note the PR for scripting support will need public access to this method

Comment (peterbarker) [reply libraries/AP_Frsky_Telem/AP_Frsky_SPort.cpp]: Fixed.

File: libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+#define ATTIANDRNG_PITCH_LIMIT      0x3FF
+#define ATTIANDRNG_PITCH_OFFSET     11
+#define ATTIANDRNG_RNGFND_OFFSET    21
+
+bool AP_Frsky_SPort_Passthrough::init()
+{
+    if (!AP_RCTelemetry::init()) {
+        return false;
+    }
+
```
Comment (yaapu) [inline libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp:61]: suggested change:
    
// do not initialize serial ports for external data
if (_use_external_data) {
  return true;
}

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp [hunk 3]
```diff
@@ -49,939 +44,42 @@ AP_Frsky_Telem::~AP_Frsky_Telem(void)
...
-            buf2[len++] = c ^ FRAME_XOR;
-        } else {
-            buf2[len++] = c;
-        }
+    AP_HAL::UARTDriver *port;
+    if ((port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_D, 0))) {
+        _backend = new AP_Frsky_D(port);
+    } else if ((port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_SPort, 0))) {
+        _backend = new AP_Frsky_SPort(port);
+    } else if ((port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_SPort_Passthrough, 0))) {
```
Comment (yaapu) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp:60]: suggested change:

    } else if (use_external_data || (port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_SPort_Passthrough, 0))) {
        _backend = new AP_Frsky_SPort_Passthrough(port, use_external_data);
    }

**Comment by yaapu** [comment]
with the above suggested changes fport telemetry works.

File: libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp:63]: I've done this one in a slightly different way- not calling your parent class' init method isn't usually a good idea.

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: Done.

**Comment by peterbarker** [comment]
@yaapu could you test again, please?  I still don't appear to have a working FPort setup here :-/  RC but no data.

**Comment by yaapu** [comment]
@peterbarker did a very quick test and fport works on r-xsr on D16 v1.x.x

**Review by tridge — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks @yaapu

**Comment by yaapu** [comment]
@peterbarker you've done the hard part!

**Pull request merged**